target/ppc/translate/fp-impl.c.inc

```diff
@@ -589,141 +589,128 @@ static void gen_mcrfs(DisasContext *ctx)
     tcg_temp_free_i64(tnew_fpscr);
 }
 
-/* mffs */
-static void gen_mffs(DisasContext *ctx)
+static void do_mffsc(int rt, TCGv_i64 t1, uint64_t set_mask,
+                     uint64_t clear_mask, uint32_t fpscr_mask)
 {
-    TCGv_i64 t0;
-    if (unlikely(!ctx->fpu_enabled)) {
-        gen_exception(ctx, POWERPC_EXCP_FPU);
-        return;
-    }
-    t0 = tcg_temp_new_i64();
+    TCGv_i64 fpscr;
+
+    fpscr = tcg_temp_new_i64();
+
     gen_reset_fpstatus();
-    tcg_gen_extu_tl_i64(t0, cpu_fpscr);
-    set_fpr(rD(ctx->opcode), t0);
-    if (unlikely(Rc(ctx->opcode))) {
-        gen_set_cr1_from_fpscr(ctx);
-    }
-    tcg_temp_free_i64(t0);
+    tcg_gen_extu_tl_i64(fpscr, cpu_fpscr);
+    tcg_gen_andi_i64(fpscr, fpscr, set_mask);
+    set_fpr(rt, fpscr);
+
+    tcg_gen_andi_i64(fpscr, fpscr, clear_mask);
+    tcg_gen_or_i64(fpscr, fpscr, t1);
+
+    gen_helper_store_fpscr(cpu_env, fpscr, tcg_constant_i32(fpscr_mask));
```

Comment:
Don't call `gen_helper_store_fpscr` if the instruction is not setting FPSCR.
(Also, we should probably resolve the TODO in this helper...)

Reply:
Sorry, I don't understand the issue with `gen_helper_store_fpscr` here. Could you give an example?

Reply:
`trans_MFFS` and `trans_MFFSL` call `do_mffsc`, but they don't set FPSCR, so the call to `gen_helper_store_fpscr` is unnecessary. You can use `set_mask`/`clear_mask` values or add another parameter to control whether the helper should be called. Or split `do_mffsc` in two methods.